[WIP] Handling DTMF tones in a steaming phone call

shahafabileah opened the pull request:
Discussed here:
https://discord.com/channels/1079125925163180093/1105907413992882238/1106005912184967248

The goal is to do an outbound call where the bot will need to navigate a phone tree by listening to options and pressing keys on the dial pad.

The first step is to have `wav` files for the DTMF tones so we can play them through the streaming conversation.  I explored several options for where to get these wav files.  The first attempts either wouldn't play through the python `wave` library, or sounded good on localhost but not through a Twilio call.  Eventually I wrote a python script `vocode/streaming/synthesizer/dtmf_audio/generate_wav_files.py` to generate the wav files in that directory using the `sox` commandline tools.  These files seem to work both on localhost and through Twilio (at least in the sense that I can hear the tone -- I haven't tried it on an actual IVR yet).

Then I added some logic into streaming_conversation.py to detect the intent to press a button from messages with a given format: `*presses 0*`.  If we detect that format, we play the sound.  Note that the bot _still_ also says "presses 0" after the tone, which is a little strange but probably harmless, and helps in the sense that we see that in the final call transcript.

At this point it works in my little simulated test:
I triggered an `OutboundCall` to my mobile phone with the `prompt_preamble`: "Your name is George. You are calling ACME Corp about order number 1 2 3 4 5. You want to reach the billing department. If you need to press a button on the phone, reply with the format '*presses 1*' and nothing more."

Here's the log output:
```
DEBUG:main:Got transcription:  Thank you for calling. If you'd like to reach billing department, please press one., confidence: 0.99072266
DEBUG:main:Checking if message is DTMF: *presses 1*
DEBUG:main:Detected DTMF: 1
DEBUG:main:Message sent: *presses 1*
DEBUG:main:Got transcription:  This is for an existing order, please press one. For new orders, please press two., confidence: 0.9995117
DEBUG:main:Checking if message is DTMF:  *presses 1*
DEBUG:main:Detected DTMF: 1
DEBUG:main:Message sent:  *presses 1*
DEBUG:main:Got transcription:  Hello. This is. How can I help you?, confidence: 0.9995117
DEBUG:main:Checking if message is DTMF: Hello, my name is George and I would like to inquire about the status of my order number 1 2 3 4 5.
DEBUG:main:Message is not DTMF
DEBUG:main:Message sent: Hello, my name is George and I would like to inquire about the status of my order number 1 2 3 4 5.
DEBUG:main:Checking if message is DTMF: Can you please provide me with an update on the billing and payment status of this order?
DEBUG:main:Message is not DTMF
DEBUG:main:Message sent: Can you please provide me with an update on the billing and payment status of this order?
```

Things that remain:
1. How do we allow people to enable/disable this feature?
2. How/should we allow people to configure what's the pattern to look for when identifying the intent to press a DTMF key?  Is it even a good approach to use a regex pattern like that or do we need something more structured/reliable?
3. Is it OK that the robot both plays the tone and still speaks "presses 1"?  If not, should we find some other way for the DTMF tones to be indicated in the transcript?

shahafabileah commented on vocode/streaming/synthesizer/base_synthesizer.py:
```diff
@@ -26,6 +26,14 @@
+    # The original audio files didn't work with the wave library, it gave this error:
+    #   wave.Error: unknown format: 65534
+    # The fix was to uncompress them using the sox commandline tool:
+    #   sox 164396__ialexs__dtmf-0.wav -b 16 dtmf-0.wav
```
I'm not sure this is the best way to convert the audio file.  I took the suggestion from here:
https://github.com/jiaaro/pydub/issues/129#issuecomment-948220528

It works in the sense that I _can_ hear the file play using the play_dtmf.py playground script.  But it may not be optimal.

ajar98 commented:
hey @shahafabileah , mind rebasing? I did make a large change to `streaming_conversation` - let me know if you need help

shahafabileah commented:
@ajar98 I rebased and updated the PR to work with the new streaming conversation setup.

I'm able to run this through a Twilio call and to hear the tones played (in addition to the speech "presses 1" or whatnot).  But as noted in [discord](https://discord.com/channels/1079125925163180093/1079125925830078466/threads/1108892560178552833), I'm not getting any audio out of my _local_ speaker when I run like that, without Twilio (though this issue seems unrelated to my DTMF change because I can't get that audio to work even with fresh code from main).

I ended up putting the logic in SynthesisResultsResultsWorker.  I'm not sure if that's the best spot.  Please advise.

And, as noted before, the following questions remain:
1. How do we allow people to enable/disable this feature?
2. How/should we allow people to configure what's the pattern to look for when identifying the intent to press a DTMF key? Is it even a good approach to use a regex pattern like that or do we need something more structured/reliable?
3. Is it OK that the robot both plays the tone and still speaks "presses 1"? If not, should we find some other way for the DTMF tones to be indicated in the transcript?

shahafabileah commented:
Update from our discussion in Discord -
I tried running the newer [streaming_conversation.py](https://github.com/vocodedev/vocode-python/blob/main/quickstarts/streaming_conversation.py) code, and it _does_ work on my laptop (mic + speakers), including the DTMF tones.

shahafabileah commented:
@ajar98 I tested this on an actual IVR system and sadly it doesn't seem to work.  So, even while I can hear the audio being played in the call (on the recording), the tone must not be an exact match, so it's not registered.

I'll try to find another way to play these tones so that they actually work.  If you have suggestions, let me know.

I know that Twilio has a way to play digits:
https://www.twilio.com/docs/voice/twiml/play#examples-2

Off hand it's not clear to me how this kind of VoiceResponse would work with the vocode setup.  It looks like the `Call` class has access to the `twilio_sid`, while the parent `StreamingConversation` class doesn't have that detail.  But the approach I took so far is to have `StreamingConversation` be the spot where we detect the _need_ to play a DTMF tone.

ajar98 commented:
tracking conversation in https://discord.com/channels/1079125925163180093/1110400160950927370

ajar98 reviewed (commented):
hey @shahafabileah can we close this since it seems like the Twilio Stream swallows up DTMF tones?

shahafabileah commented:
Sure.  Closing it...
